### DISCUSSION TITLE
Performance issue when `persist` was called inside a for loop

### DISCUSSION THREAD

**Discussion by haoxi911** [mikro-orm/mikro-orm#3179] [Q&A] [answered]
I noticed a performance issue with the following code, it loops about 200 times, in each loop it will run a query, then create a new record if needed, otherwise update the existing record.

I am using Mikro-ORM with SQLite. On my MacBook Pro 2019 running macOS 12, when the database file is stored on SSD, the code below took about 1.5s, when the database file is stored on a FAT32 flash drive, it took about 9s. It looks worse on Windows, when I test the same code on Windows 11 with the same FAT32 flash drive, the code below took more than 30s. Please note it is a simple for loop with only 200 records.

**Example Code (Slow)**
```
const repository = em.fork().getRepository(...);

// loop for ~200 times
for (const asset of assets) {

  // obtain raw data object.
  const data = wrap(asset).toObject();

  // create or update
  let item = await repository.findOne({primaryKey: asset.key});
  if (!item) {
    item = repository.create(data);
    await repository.persist(item);
  } else {
    wrap(item).assign(data);
  }

}

await repository.flush();
```

I found a simple fix, instead of running `persist` inside the for loop, I can save the new created entities into an array, and flush all of them at the end of the loop. The code below took only 500ms on Windows 11 with FAT32 flash drive.

**Example Code (Fast)**
```
const repository = em.fork().getRepository(...);
const items = [];

// loop for ~200 times
for (const asset of assets) {

  // obtain raw data object.
  const data = wrap(asset).toObject();

  // create or update
  let item = await repository.findOne({primaryKey: asset.key});
  if (!item) {
    item = repository.create(data);
    items.push(item);
  } else {
    wrap(item).assign(data);
  }

}

await repository.persistAndFlush(items);
```

**Questions**
1. Why `persist` in the for loop would slow down the app?
2. In my 2nd code block, if some records were matched during `findOne` and their attributes were updated in `wrap(item).assign(data)`, will `persistAndFlush` be able to save these changes? If not, should I pass all items into `persistAndFlush`?

**Versions**

| Dependency | Version |
| - | - |
| node | v16.14.0 |
| typescript | v4.6.3 |
| mikro-orm | v5.1.4 |
| your-driver | sqlite |

**Comment by B4nan** [comment, marked as answer]
Please provide complete reproduction, this will be definitely depenent on the complexity of your entities and the particular data you pass in. Note that persist is sync method, you should not await it. Calling em.persist wont cascade so it should be fast.

**Reply by haoxi911** [reply]
Hey Martin,

Thank you so much for getting back to me.

I have removed `await` before `persist` calls, I also confirmed the entity I am using doesn't have any cascading or foreign keys, it is a simple and independent entity like:
```
{
  localPath: 'C:\\Users\\Kevin\\Pictures\\iPhone\\IMG_1311.JPG',
  fileSize: 1951531,
  cloudState: 0,
  assetType: 1,
  sha1Hash: null,
  createdDate: 2021-04-21T15:36:40.000Z,
  modifiedDate: 2021-04-01T04:12:07.000Z,
  takenDate: null,
  thumbnail: null,
  driveUuid: '4c6334df-0a98-41a6-a2c6-3c4d316d7f08'
}
```

I am thinking about looking into the source code of `persist` method and add some `console.time` to narrow it down, I shall let you know if I have any findings. 

By the way, would you mind take a look at my 2nd question above? Will `persistAndFlush` save any changes to the entities returned from `find`, or do I need to explicitly pass all changed entities as an array into `persistAndFlush`.

Thank you!

**Reply by B4nan** [reply]
>  it is a simple and independent entity like:

Maybe its something else, thats why I'd like to see complete reproduction, including data seeding. Btw here is a test case calling persist 1000 times on rather complex entity, works fine (whole test case locally tooks around 400ms on my m1):

https://github.com/mikro-orm/mikro-orm/blob/master/tests/EntityManager.postgre.test.ts#L1877

> Will persistAndFlush save any changes to the entities returned from find, or do I need to explicitly pass all changed entities as an array into persistAndFlush.

Flush is what saves to database, and it saves what is in the context. You dont need to persist managed entities at all, persist call just makes the EM aware of your entity, if its managed, its already aware of it.

**Reply by haoxi911** [reply]
I understand, I shall make a minimal sample later.

I looked into in the source code a little bit and found this [line](https://github.com/mikro-orm/mikro-orm/blob/eb037598ea295519f9d17ce3b08b4562dca04b4a/packages/core/src/unit-of-work/UnitOfWork.ts#L252), if I commented it out then the speed is back to normal. Maybe it makes a difference as I am using file path as primary key and also my test set has very similar primary keys?

The file path (primary key) of my 200 records look like:
```
C:\\Users\\Kevin\\Pictures\\iPhone\\IMG_1001.JPG
C:\\Users\\Kevin\\Pictures\\iPhone\\IMG_1002.JPG
...
C:\\Users\\Kevin\\Pictures\\iPhone\\IMG_1200.JPG
```

On the other hand, the issue only occur when I call `findOne => create => persist` in the loop, if I remove `findOne` then running `create => persist` 200 times only takes ms.

**Reply by B4nan** [reply]
> I looked into in the source code a little bit and found this line

Well, new entities needs to be part of persist stack, if you comment that line, most probably many things wont get persisted at all. So it looks fast because it probably does not do any queries (maybe only the update ones).

I'll wait for the repro, it sounds weird for sure, but as you can see in the test I linked, persisting in a loop on its own should not matter much. Maybe there is some cascading happening because of the `em.create()` but it sounds almost impossible that just the `em.persist()` would be slow (it can cause the slowness for sure, but that call on its own will be fast).

edit: the reason why its slow is probably the context that gets bigger and bigger on each iteration, so it would rather point to `em.create()` which actually checks the context, the `em.persist()` will just check the persist stack, which is a `Set` of objects, the look up there is extremely fast.

**Reply by haoxi911** [reply]
Here is a minimal sample project [mikro-test.zip](https://github.com/mikro-orm/mikro-orm/files/8839189/mikro-test.zip)

What happened to me was, after `npm install` and `npx tsc`, running `node index.js` gave me the following output:

<img src="https://user-images.githubusercontent.com/12502170/172031654-9d50b3c5-8e27-4ea8-ad93-c8e6364722c1.png" width="50%">

Honestly, I believe the issue is Windows only (I am running Windows 11 Pro), when I made this minimal project, I wrote the code on macOS but was not able to reproduce the issue, the 200 for-loop took ~500ms on macOS. However, the same code on my Windows 11 took about 15s, so weird.

I am not sure how to run your unit test on Windows, or if you have access to a Windows 11 machine. Maybe the issue will reproduce with your unit test too.

**Reply by B4nan** [reply]
Can't really help with windows issues, the repro works fine on my end and it is pretty much the same as the test case I already presented. Also I can't really understand how this could be platform specific, sounds more like you have something broken on your windows machine or its just generally much slower.

**Reply by haoxi911** [reply]
No problem, I agree it is a very weird issue. I will keep looking into it and let you know if I find anything. Meanwhile, I will close the issue. Thanks for your help!
